tests/v2/validation/charts/gatekeeper.go

```diff
@@ -0,0 +1,112 @@
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	management "github.com/rancher/rancher/tests/framework/clients/rancher/generated/management/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	// Project that charts are installed in
+	gatekeeperProjectName = "gatekeeper-project"
+)
+
+var Constraint = schema.GroupVersionResource{
+	Group:    "constraints.gatekeeper.sh",
+	Version:  "v1beta1",
+	Resource: "k8srequiredlabels",
+}
+
+var Namespaces = schema.GroupVersionResource{
+	Group:    "",
+	Version:  "v1",
+	Resource: "namespaces",
+}
+
+type Status struct {
+	AuditTimestamp  string
+	ByPod           interface{}
+	TotalViolations int64
+	Violations      []interface{}
+}
+type Items struct {
+	ApiVersion string
+	Kind       string
+	Metadata   interface{}
+	Spec       interface{}
+	Status     Status
+}
+
+// ConstraintResponse is the data structure that is used to extract data about the gatekeeper constraint created in the test
+// It contains the Items and Status structs, which are used for the same purpose
+// anything that isn't being used in the test is declared as a string or an interface
+type ConstraintResponse struct {
+	ApiVersion string
+	Items      []Items
+	Kind       string
+	Metadata   interface{}
+}
+
+// GetUnstructuredList helper function that returns an unstructured list of data from a cluster resource
+func getUnstructuredList(client *rancher.Client, project *management.Project, schema schema.GroupVersionResource) *unstructured.UnstructuredList {
+	dynamicClient, err := client.GetDownStreamClusterClient(project.ClusterID)
+	if err != nil {
+		fmt.Println("error: ", err)
+	}
+
+	unstructured := dynamicClient.Resource(schema).Namespace("")
+
+	unstructuredList, err := unstructured.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		fmt.Println("error: ", err)
+	}
+
+	return unstructuredList
+}
+
+// parseConstraintList converts an Unstructed List into a Constraint response by marshaling it into json, then unmarshaling the json into the Constraint respionse struct, allowing access to deeply nested fields
+func parseConstraintList(list *unstructured.UnstructuredList) *ConstraintResponse {
+	jsonConstraint, err := list.MarshalJSON()
+	if err != nil {
+		fmt.Println("error: ", err)
+	}
+
+	var parsedConstraint ConstraintResponse
+	err = json.Unmarshal([]byte(jsonConstraint), &parsedConstraint)
+	if err != nil {
+		fmt.Println("error: ", err)
+	}
+
+	return &parsedConstraint
+
+}
+
+func getAuditTimestamp(client *rancher.Client, project *management.Project) {
+	// sleep until the first audit finishes running.
+	// AuditTimestamp will be empty string until first audit finishes
+	// audit runs every 60 seconds, plus an arbitrary amount of time to set up the audit pod and for the audit itself
+	counter := 0
+	auditTime := ""
+	for auditTime == "" && counter < 5 {
```

Comment:
I see what you are saying now I recommend using this polling package `k8s.io/apimachinery/pkg/util/wait` and do something like this instead 

```
wait.Poll(500*time.Millisecond, 2*time.Minute, func() (done bool, err error) {
                auditList := getUnstructuredList(client, project, Constraint)

 		// parse it so that we can extract individual values
 		parsedAuditList := parseConstraintList(auditList)

 		// extract the timestamp of the last constraint audit
 		auditTime = parsedAuditList.Items[0].Status.AuditTimestamp
		if auditTime == "" {
                    return false, nil  
               }
		
		return true, nil
	})
```